### ISSUE TITLE
Race condition closing stream just after writing data

### ISSUE THREAD

**Issue by paralin** [lucas-clemente/quic-go#3472]
I'm encountering an issue where, when writing to a Stream:

```go
	qstream.Write([]byte{"data"})
	qstream.Close()
```

Close() indicates "close for writing" - send the data written before, then EOF.

However, the "data" packet is never sent if the Close( ) comes immediately after Write.

If I add a `time.Sleep(time.Millisecond)` between the `Write` and `Close`, it works.

The other end receives a "close" instantly and no data packet.

Perhaps there's a race condition in the code where the last packet is not sent if Close is called instantly after Write?

**Comment by marten-seemann** [comment]
The stream is carefully designed such that this shouldn't happen. Can you point to exactly where you think the race happens?

**Comment by paralin** [comment]
@marten-seemann I thought as much; however once I could reproduce the issue I posted here.

That said... I checked into my Write() call stack a bit further, and discovered that probably the Quic Write() was returning `bytesWritten < len(data)` (as designed) and I was (incorrectly) not repeatedly calling Write until bytesWritten == len(data).

Closing this since after I fixed that, I'm not able to reproduce. Thanks.
